src/IB/IBFEMethod.cpp

```diff
@@ -1931,29 +2486,80 @@ IBFEMethod::computeStressNormalization(PetscVector<double>& Phi_vec,
+                        Phi += J * P;
+                    }
                 }
 
                 // Add the boundary forces to the right-hand-side vector.
                 for (unsigned int i = 0; i < n_basis; ++i)
                 {
-                    Phi_rhs_e(i) += PENALTY * Phi * phi_face[i][qp] * JxW_face[qp];
+                    if ((solver_flag.compare(CG_PHI_SOLVER_NAME) == 0) || (solver_flag.compare(CG_DIFFUSION_PHI_SOLVER_NAME) == 0))
+                    {
+                        Phi_rhs_e(i) += cg_penalty * Phi * phi_face[i][qp] * JxW_face[qp];
+                    }
+                    else if (solver_flag.compare(IPDG_PHI_SOLVER_NAME) == 0)
+                    {
+                        Phi_rhs_e(i) += diffusion * JxW_face[qp] * Phi * ipdg_jump0_penalty / h0_elem * phi_face[i][qp];
+                        Phi_rhs_e(i) -= diffusion * JxW_face[qp] * dphi_face[i][qp] * (Phi * normal_face[qp]);
+                    }
+                    else // default solver to CG
+                    {
+                        Phi_rhs_e(i) += cg_penalty * Phi * phi_face[i][qp] * JxW_face[qp];
+                    }
                 }
             }
+        }
+
+        if (solver_flag.compare(CG_DIFFUSION_PHI_SOLVER_NAME) == 0)
+        {
+            for (unsigned int qp = 0; qp < qrule->n_points(); qp++)
+            {
+                Number Phi_old = 0.0;
+                Gradient grad_Phi_old;
 
+                // get values of solution and its gradient at the previous timestep
+                for (unsigned int l = 0; l < phi.size(); l++)
+                {
+                    Phi_old += phi[l][qp] * Phi_system.old_solution(Phi_dof_indices[l]);
+                    grad_Phi_old.add_scaled(dphi[l][qp], Phi_system.old_solution(Phi_dof_indices[l]));
+                }
+
+                for (unsigned int i = 0; i < phi.size(); i++)
+                {
+                    // for timestepping
+                    Phi_rhs_e(i) +=
+                        JxW[qp] * (Phi_old * phi[i][qp] - diffusion * 0.5 * dt * grad_Phi_old * dphi[i][qp]);
+                }
+            }
+        }
+
+        if ((solver_flag.compare(CG_PHI_SOLVER_NAME) == 0) || (solver_flag.compare(CG_DIFFUSION_PHI_SOLVER_NAME) == 0))
+        {
             // Apply constraints (e.g., enforce periodic boundary conditions)
             // and add the elemental contributions to the global vector.
             Phi_dof_map.constrain_element_vector(Phi_rhs_e, Phi_dof_indices);
-            Phi_rhs_vec->add_vector(Phi_rhs_e, Phi_dof_indices);
         }
+
+        Phi_rhs_vec->add_vector(Phi_rhs_e, Phi_dof_indices);
     }
 
     // Solve for Phi.
     Phi_rhs_vec->close();
     Phi_system.solve();
     Phi_system.solution->close();
     Phi_system.solution->localize(Phi_vec);
-    Phi_dof_map.enforce_constraints_exactly(Phi_system, &Phi_vec);
+
+    if ((solver_flag.compare(CG_PHI_SOLVER_NAME) == 0) || (solver_flag.compare(CG_DIFFUSION_PHI_SOLVER_NAME) == 0))
```

Comment:
Can you add a brief comment explaining why you do not need to enforce constraints for other solver choices?